### PR TITLE
Gtag ecommerce tracking improvements

### LINKED ISSUE #37348: GTag not send transaction prices and totals.
### Preconditions and environment

- Magento version: >= 2.4.5-p1, 2.4.5-p2, 2.4.6

### Steps to reproduce

1. Enable Gtag (Store - Configuration - Sales - Google API - GTag - Enable and add Gtag Key
2. Add product to cart and finish an order.
3. On success page you need capture OrderTrackingData which was send to Google Tag.

### Expected result

Currency need to be part of the order, not part of the "all" orders provided. Reference from Google Analytics docs: https://developers.google.com/analytics/devguides/collection/ga4/ecommerce?client_type=gtag#make_a_purchase_or_issue_a_refund

image

### Actual result

image

### Additional information

_No response_

### Release note

_No response_

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by Bashev**
### Related Pull Requests
https://github.com/magento-gl/magento2-infrastructure/pull/26

### Fixed Issues (if relevant)

1. Fixes magento/magento2#37348

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by Bashev** [comment]
@magento run all tests

**Comment by Bashev** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

File: app/code/Magento/GoogleGtag/Block/Ga.php [hunk 1]
```diff
@@ -63,13 +70,16 @@ public function __construct(
         SerializerInterface $serializer,
         SearchCriteriaBuilder $searchCriteriaBuilder,
         OrderRepositoryInterface $orderRepository,
+        ProductRepositoryInterface $productRepository,
```
Comment (engcom-Hotel) [inline app/code/Magento/GoogleGtag/Block/Ga.php]: I think we should make it null be default in accordance to backward compatibility.

File: app/code/Magento/GoogleGtag/Block/Ga.php [hunk 2]
```diff
@@ -63,13 +70,16 @@ public function __construct(
...
         array $data = []
     ) {
         $this->googleGtagConfig = $googleGtagConfig;
         $this->cookieHelper = $cookieHelper;
         $this->serializer = $serializer;
         $this->orderRepository = $orderRepository;
         $this->searchCriteriaBuilder = $searchCriteriaBuilder;
+        $this->productRepository = $productRepository;
```
Comment (engcom-Hotel) [inline app/code/Magento/GoogleGtag/Block/Ga.php]: And use ObjectManager to initialize $this->productRepository

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @Bashev,

Thanks for the contribution!

Please review my comment below and fix the build failures below.

Thanks

**Comment by Bashev** [comment]
@magento run all tests

**Comment by Bashev** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Magento Health Index, Static Tests, Unit Tests

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @Bashev,

Thanks for working on the review comments. Please refer to the build failures as well. It seems the failures are due to PR changes.

Thanks

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Integration Tests,Functional Tests B2B

**Comment by Bashev** [comment]
@magento run Integration Tests, Functional Tests B2B

**Comment by Bashev** [comment]
@magento run Functional Tests B2B

**Comment by Bashev** [comment]
@magento run Sample Data Tests B2B, Sample Data Tests CE, Sample Data Tests EE

**Comment by Bashev** [comment]
@magento run Sample Data Tests B2B, Sample Data Tests CE, Sample Data Tests EE

**Review by engcom-Hotel — APPROVED** [review]
The changes look good to me. For the `Magento Health Index` failure we need to create an approval JIRA later.

Hence approving this PR.

**Comment by Bashev** [comment]
@engcom-Delta fix of currency is initiated before 2.4.6-p2, but improvements are still valid.
1. Brand is added in the data.
2. Coupon code from the order is also added in the data.

**Comment by engcom-Delta** [comment]
Hi @Bashev , 

We have tried with brand and coupon code as well but this is not reproducible.

Can you please clarify the steps to reproduce again.


Screenshot 2023-10-11 at 7 43 02 PM


Screenshot 2023-10-11 at 8 06 36 PM


Added screenshots for reference.

**Comment by engcom-Delta** [comment]
Hi @Bashev ,

Thank you for your contribution. As mentioned here https://github.com/magento/magento2/pull/37504#issuecomment-1757838046 , we have tried to reproduce the issue and we couldn't able to reproduce it. We are able to see currency here.

Please confirm once.

**Comment by Bashev** [comment]
Hi @engcom-Delta,
below are described steps and results of my test.

1. Add brand to products (or just for this one which will use in the purchase).
2. Create Cart pricing rule with specific coupon.

2. Go to webshop
3. Add products from point 1. 
4. Go to checkout
5. Start checkout process - fill shipping data and choose shipping method, confirm and go to payment page.
6. Add coupon code from point 2.
7. Finilize order.

Screenshot 1. (Ordered product with Discount)
image

Screenshot 2. Debug from Google Tag Manager - DataLayer with pushed data with brand and applied coupon.
image

Screenshot 3. Google Analytics 4 (Debug mode) - Parsed parameters from the data layer with Brand and applied coupon.
image

**Comment by engcom-Delta** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Functional Tests CE

**Comment by engcom-Echo** [comment]
### :heavy_check_mark: QA Passed

**Preconditions:**
- Install fresh Magento **`2.4-develop`**

**Manual testing scenario:**

1. We have follow steps from [comment](https://github.com/magento/magento2/issues/37348#issue-1659457235)
2. Create Coupon and Use same coupon.

**Before:** :heavy_multiplication_x:  Coupon and Discount  data  is not present in gtag data
Screenshot 2023-12-13 at 6 26 35 PM

**After:** :heavy_check_mark:  Coupon and Discount is present in gtag data
Screenshot 2023-12-13 at 6 32 04 PM

Builds are failing, Hence moving to `Extended Testing`

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Integration Tests,Functional Tests B2B

**Comment by engcom-Echo** [comment]
Functional Tests B2B builds are green and Static Tests are known failure.
Hence moving it to `Merge In Progress`

**Comment by engcom-Charlie** [comment]
Hi @Bashev,

As per the static test failure:
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/37504/2e4c1e56737a74d963734ed5da13d4d0/Statics/console-error-logs.html

The view layer changes have been detected in the GoogleGtag module. The GoogleGtagGraphQl module is expected to be updated to reflect these changes. The test failure can be ignored if the changes can not be covered with GraphQL API.

In order to proceed further we need to green build so can you please look into this and update the PR accordingly. Moving this to Changes Requested now.

Thank you!

image

**Comment by engcom-Charlie** [comment]
Hi @Bashev,

As mentioned in above [comment](https://github.com/magento/magento2/pull/37504#issuecomment-1913963110), static test failure need to be fix in order to proceed further on this PR. Currently closing this PR, please feel free to reopen for any updates from your end.

Thank you!

**Pull request closed without merging**